taar/recommenders/utils.py

```diff
@@ -24,21 +28,44 @@ def fetch_json(uri):
     return r.json()
 
 
+def get_s3_cache_filename(s3_bucket, s3_key):
+    return os.path.join(gettempdir(),
+                        '_'.join([s3_bucket, s3_key]).replace('/', '_'))
+
+
 def get_s3_json_content(s3_bucket, s3_key):
     """Download and parse a json file stored on AWS S3.
 
     The file is downloaded and then cached for future use.
     """
-    local_filename = '_'.join([s3_bucket, s3_key]).replace('/', '_')
-    local_path = os.path.join(gettempdir(), local_filename)
+    local_path = get_s3_cache_filename(s3_bucket, s3_key)
 
     if not os.path.exists(local_path):
-        with open(local_path, 'wb') as data:
+        # Use NamedTemporaryFile, so that the file gets removed.
+        with NamedTemporaryFile() as temp_file:
             try:
                 s3 = boto3.client('s3')
-                s3.download_fileobj(s3_bucket, s3_key, data)
+                s3.download_fileobj(s3_bucket, s3_key, temp_file)
+                # Flush the file.
+                temp_file.flush()
             except ClientError:
+                logger.exception("Failed to download from S3", extra={
+                    "bucket": s3_bucket,
+                    "key": s3_key})
                 return None
 
-    with open(local_path, 'r') as data:
-        return json.loads(data.read())
+            with open(local_path, 'wb') as data:
+                temp_file.seek(0)
+                shutil.copyfileobj(temp_file, data)
+
+    # It can happen to have corrupted files. Account for the
+    # sad reality of life.
+    try:
+        with open(local_path, 'r') as data:
+            return json.loads(data.read())
+    except:
```

Comment:
What about catching a ValueError? That's what json.loads throws when the content cannot be deserialized